### DISCUSSION TITLE
3D mapping and topology

### DISCUSSION THREAD

**Discussion by turepalsson** [osmlab/osm-data-model#38] [General]
How do we handle points with the same (lat,lon) coordinates but different levels?

For example, in this (somewhat sketchy) multi-storey car park, where the red line represents a `highway=service` access road, the two points circled in blue have the same positions projected on the ground plane, but different elevations. For the horizontal bits of the way, we could use `level` tags, but how to handle the slopes, which have different levels at each end?

One possibility would be something like `start_level=1`,`end_level=2` or even `level=1;2`, though I’m not super fond of tags where the value depends on the orientation of the way.

![P-hus](https://user-images.githubusercontent.com/14193668/211151414-b6e90a27-77cb-4584-b80c-4b30af3c47c8.jpg)

**Comment by tordanik** [comment]
Currently, the values in a `level` tag are considered unordered. Using established tags, the situation could be made unambiguous by adding `incline=up` or `incline=down`.

Certain use cases such as calculating a route would have to start looking at layer/level information to still produce correct results after the change.

**Reply by turepalsson** [reply]
Oops, I wasn’t even aware that `levels` could have multiple tags today. =)

**Reply by flacombe** [reply]
With all due respect I think it's a pretty bad idea to solve core data model issues with tagging.

Tagging is high level conceptual model and shouldn't have any influence on lower layers.
This would be inconsistent with area type definition where it is clearly intended to remove tags from decision making in area/polygon guessing.

**Reply by turepalsson** [reply]
I agree that solving data model limitations with tags does not feel good... after all, this is how we got into the racetrack-or-area mess that we are now trying to get out of!

**Comment by 1ec5** [comment]
This gets tricky when vertices of a single way, such as a building way, exist on multiple levels. The simple indoor mapping scheme has affordances for repeating, say, a door on multiple levels. However, this doesn’t address one situation I often run into, that there’s a statue or flagpole mounted on the wall directly over a doorway, requiring a different set of tags that would be weird to mix. It isn’t great to have to overlap two nodes of the same way to express this fact. Because of validator warnings, mappers tend to offset one of the features slightly, but it results in an imbalanced 3D rendering.

**Reply by tordanik** [reply]
This is very true, and it emphasizes that we already have issues with vertically stacked nodes in the current data model. Today, it is kind of expected that window and door nodes are part of the building outline. However, that means that vertically stacked doors and windows would result in zero-length way segments, which mappers attempt to avoid by offsetting the node by a small amount. (Which indeed isn't ideal because an offset of a few cm per level can absolutely be visible in a 3D rendering, because it's really fiddly to do with any existing editor, and because it's not inutitive that you need to do it in the first place.)

So today's situation isn't exactly ideal either. In a fashion, the proposed change would perhaps _lessen_ this particular aspect of the problem by eliminating the concept of nodes being part of a way from the data model, and therefore force us to use other criteria for deciding which building a door/window/flagpole node belongs to. The stacked nodes could then all have the same coordinates as one of the way's points, but this would not create additional way segments.

**Reply by mmd-osm** [reply]
I think it would be helpful to have a concrete example how this translates into an OSM XML representation, both using the current, and a future version, assuming a way format as follows:

```xml
  <way id="273615">
    <nd lat="50.9700346" lon="7.9897718"/>
    <nd lat="50.9700357" lon="7.9897722"/>
...
    <tag k="tag" v="value"/>
...    
  </way>
```

According to my understanding of the proposed data model changes,  identical coordinates in different ways would always mean both points are topologically connected (which is an issue if they are on different levels).

> force us to use other criteria for deciding which building a door/window/flagpole node belongs to

Do you have some examples what those could be?

In the end, what I really want to find out is if an automated translation from 0.6 to 0.X is still feasible in case of indoor or 3D mapping.

From earlier discussions  I have learnt that two nodes sharing the same lat/lon values is usually seen as a rare special case (and mostly incorrect). If this is going to change in significant ways due to 3D mapping, then this needs some further consideration.

On this topic, the data model study suggests: "_But detailed indoor tagging is a minority use for OSM and we can come up with a solution for this special case using level tags or so._"

**Reply by bryceco** [reply]
I was thinking about this today while trying to map a flagpole that contains two flags, while there is only room to tag one flag per node. One possible hack would be to create a relation that contains only the flagpole node as a member, set the role to something like "is_also" and then put the tags for the second flag on the relation. You could do something similar for points that span multiple floors.

**Reply by 1ec5** [reply]
It’s common to map a single `man_made=flagpole` node with semicolon-delimited values for `flag:wikidata` etc. After all, what’s being mapped is not the flags themselves (which are kind of ephemeral) but rather the flagpole that supports them. The identities of the flags are ancillary information.

While this approach works reasonably well for multiple flags on flagpoles, it doesn’t help as much for multiple signs assembled on the same signpost. The current [`traffic_sign`](https://wiki.openstreetmap.org/wiki/Key:traffic_sign) documentation recommends semicolon- or comma-delimited values. However, this is different than flagpoles because it’s the signs themselves that are being mapped as features in their own right. A lot of attributes can differ between the signs within an assembly, but the data model doesn’t allow us to nest an array of key-value tables inside a single tag. I would prefer to [map the signs as individual features](https://wiki.openstreetmap.org/wiki/Talk:Key:traffic_sign#Multiple_features_for_multiple_signs) to allow for more detail about each sign, but some way of associating the features would be beneficial.

**Reply by 1ec5** [reply]
> One possible hack would be to create a relation that contains only the flagpole node as a member, set the role to something like "is_also" and then put the tags for the second flag on the relation.

At one point, I considered proposing a scheme for street parking restrictions that relied on multiple relations containing the same way as a member. However, in the current data model, a limitation of this approach is that a feature’s relation memberships are inherently unordered. You’d need to rely on `layer`-like tags on each relation to rank the memberships, which could get out of sync pretty easily.

**Reply by bryceco** [reply]
In your street parking example why do you need to rank the relations? At least in the multistory building case it's necessary to have layer/level tags anyhow.

**Reply by 1ec5** [reply]
Only because some restrictions could override others – think exceptions that only apply to certain times, or under certain conditions, or that extend other restrictions. We’re currently handling this by having mappers interpret the signs and distribute the details of a sign across multiple multivalued `*:conditional` tags. This gets really confusing in some cities in the U.S., where streets typically accrue partially redundant restrictions over time.

**Reply by 1ec5** [reply]
> However, in the current data model, a limitation of this approach is that a feature’s relation memberships are inherently unordered.

Do folks see this as a major concern? It might also have something to do with the prevalence of `subarea` relation roles, which are unfortunate for how they create reference loops.

**Comment by turepalsson** [comment]
(I suppose the obvious solution is to store Z coordinates as well... However, that means either increasing database space by almost 50% or complicating things by having mixed 2D and 3D coordinates. Also, while consumer-grade GPS produces 2D coords which are perfectly usable for everything except extreme large-scale mapping, the Z coordinate is a whole different ball game, so this seems like a non-starter.)

**Reply by bryceco** [reply]
I agree that adding a z-coordinate to every point is overkill, but we could add 3-d variants of ways and areas where only those variants have 3-d points, and everything else keeps the existing 2-d points.

**Reply by flacombe** [reply]
How all of this can be edited with 2d editors, from the very first 3d point?
